gematria/datasets/exegesis_benchmark_lib.cc

```diff
@@ -261,6 +263,69 @@ Expected<BenchmarkCode> ExegesisBenchmark::parseJSONBlock(
   return BenchCode;
 }
 
+Expected<BenchmarkCode> ExegesisBenchmark::processAnnotatedBlock(
+    std::string_view BlockHex, const BlockAnnotations &Annotations) {
+  std::optional<std::vector<uint8_t>> BytesOr =
+      gematria::ParseHexString(BlockHex);
+
+  if (!BytesOr.has_value())
+    return llvm::make_error<StringError>(
+        errc::invalid_argument,
+        "Malformed basic block: invalid hex value " + BlockHex);
+
+  Expected<std::vector<gematria::DisassembledInstruction>> InstructionsOrErr =
+      gematria::DisassembleAllInstructions(
+          *LLVMMCDisassembler, ExegesisState.getInstrInfo(),
+          ExegesisState.getRegInfo(), ExegesisState.getSubtargetInfo(),
+          *LLVMMCInstPrinter, 0, *BytesOr);
+
+  if (!InstructionsOrErr) return InstructionsOrErr.takeError();
+
+  BenchmarkCode BenchmarkConfiguration;
+
+  std::vector<MCInst> Instructions;
+  Instructions.reserve(InstructionsOrErr->size());
+
+  for (const gematria::DisassembledInstruction &Instruction :
+       *InstructionsOrErr)
+    Instructions.push_back(Instruction.mc_inst);
+
+  BenchmarkConfiguration.Key.Instructions = std::move(Instructions);
+
+  BenchmarkConfiguration.Key.RegisterInitialValues.reserve(
+      Annotations.initial_regs.size());
+
+  for (const RegisterAndValue &RegisterValue : Annotations.initial_regs) {
+    struct RegisterValue ValueToAdd = {
```

Comment:
I believe the `struct` tag is unnecessary here

Reply:
It's needed here as `RegisterValue` is just a normal struct rather than being `typedef`ed or being a `enum class`. Might be something that should be fixed upstream though...

Reply:
This is only the case in C, not C++. C++ doesn't have tag namespaces. The struct tag is accepted in this context, but only for C compatibility.

Reply:
Ah.

I was getting a compile error when trying to get that to work, but looks like I should've read the whole thing:
```
gematria/datasets/exegesis_benchmark_lib.cc:299:5: error: must use 'struct' tag to refer to type 'RegisterValue' in this scope
    RegisterValue ValueToAdd = {
    ^
    struct 
gematria/datasets/exegesis_benchmark_lib.cc:298:32: note: struct 'RegisterValue' is hidden by a non-type declaration of 'RegisterValue' here
  for (const RegisterAndValue &RegisterValue : Annotations.initial_regs) {
```

It was only necessary because of the variable already named `RegisterValue`.

Reply:
Huh, TIL something new about C++ shadowing. I thought `struct` was simply ignored in that context, I didn't know it could be used to disambiguate between types and locals that shadow the same name. In C types (whether in the tag namespaces or the typedef namespace) are in an entirely different namespace from variables, so this works very differently.

Some experiments: https://godbolt.org/z/qK65xeoq6
Pretty weird stuff :)

Reply:
Interesting. Today I learned about C++ shadowing. Thanks for digging into that and showing your experiments!